File: android/src/main/kotlin/dev/note11/flutter_naver_map/flutter_naver_map/model/base/NPoint.kt
package dev.note11.flutter_naver_map.flutter_naver_map.model.base

import dev.note11.flutter_naver_map.flutter_naver_map.converter.DefaultTypeConverter.asDouble
import dev.note11.flutter_naver_map.flutter_naver_map.converter.DefaultTypeConverter.asMap
import dev.note11.flutter_naver_map.flutter_naver_map.util.CalcUtil
import dev.note11.flutter_naver_map.flutter_naver_map.util.DisplayUtil

internal data class NPoint(
    val x: Double,
    val y: Double,
) {
    companion object {
        fun fromMessageable(args: Any): NPoint = args.asMap().let { map ->
            NPoint(map["x"]!!.asDouble(), map["y"]!!.asDouble())
        }

        fun fromPointF(pointF: android.graphics.PointF): NPoint {
            val x = CalcUtil.float32To64(pointF.x)
            val y = CalcUtil.float32To64(pointF.y)

            return NPoint(x, y)
        }

        fun fromPointFWithPx(pointF: android.graphics.PointF): NPoint {
            val x = DisplayUtil.pxToDp(pointF.x)
            val y = DisplayUtil.pxToDp(pointF.y)

            return NPoint(x, y)
        }
    }

    fun toPointF(): android.graphics.PointF = android.graphics.PointF(x.toFloat(), y.toFloat())

    fun toPointFWithPx(): android.graphics.PointF =
        android.graphics.PointF(DisplayUtil.dpToPxF(x), DisplayUtil.dpToPxF(y))

    fun toMessageable(): Map<String, Double> = mapOf("x" to x, "y" to y)
}

File: ios/flutter_naver_map/Sources/flutter_naver_map/model/flutter_default_custom/NPoint.swift
import CoreGraphics

internal struct NPoint {
    let x, y: Double

    var cgPoint: CGPoint {
        CGPoint.init(x: x, y: y)
    }

    func toMessageable() -> Dictionary<String, Any> {
        [
            "x": x,
            "y": y
        ]
    }

    static func fromMessageable(_ args: Any) -> NPoint {
        let d = asDict(args, valueCaster: asDouble)
        return NPoint(x: d["x"]!, y: d["y"]!)
    }

    static func fromCGPointWithDisplay(_ cgPoint: CGPoint) -> NPoint {
        NPoint(x: cgPoint.x, y: cgPoint.y)
    }

    static func fromCGPointWithOutDisplay(_ cgPoint: CGPoint) -> NPoint {
        NPoint(x: CalcUtil.float32To64(f32: cgPoint.x), y: CalcUtil.float32To64(f32: cgPoint.y))
    }
}
